go-controller/pkg/node/port_claim.go

```diff
@@ -120,61 +150,66 @@ func updateServicePortClaim(oldSvc, newSvc *kapi.Service) []error {
-	klog.V(5).Infof("Opening socket for service: %s/%s and port: %v", svc.Namespace, svc.Name, port)
-	var socket activeSocket
-	var socketError error
+func (p *portClaimWatcher) open(desc string, ip string, port int32, protocol kapi.Protocol, svc *kapi.Service) error {
+	klog.V(5).Infof("Opening socket for service: %s/%s, port: %v and protocol %s", svc.Namespace, svc.Name, port, protocol)
+
+	var localPort *utilnet.LocalPort
+	var portError error
 	switch protocol {
 	case kapi.ProtocolTCP:
-		listener, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
-		if err != nil {
-			socketError = err
-			break
-		}
-		socket = listener
+		localPort, portError = utilnet.NewLocalPort(desc, ip, "", int(port), utilnet.TCP)
 	case kapi.ProtocolUDP:
-		addr, err := net.ResolveUDPAddr("udp", fmt.Sprintf(":%d", port))
-		if err != nil {
-			socketError = err
-			break
-		}
-		conn, err := net.ListenUDP("udp", addr)
-		if err != nil {
-			socketError = err
-			break
-		}
-		socket = conn
+		localPort, portError = utilnet.NewLocalPort(desc, ip, "", int(port), utilnet.UDP)
 	case kapi.ProtocolSCTP:
 		// Do not open ports for SCTP, ref: https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0015-20180614-SCTP-support.md#the-solution-in-the-kubernetes-sctp-support-implementation
 		return nil
 	default:
-		socketError = fmt.Errorf("unknown protocol %q", protocol)
+		portError = fmt.Errorf("unknown protocol %q", protocol)
 	}
-	if socketError != nil {
-		p.emitPortClaimEvent(svc, port, socketError)
-		return socketError
+	if portError != nil {
+		p.emitPortClaimEvent(svc, port, portError)
+		return portError
 	}
+	klog.V(5).Infof("Opening socket for LocalPort %v", localPort)
 	p.activeSocketsLock.Lock()
 	defer p.activeSocketsLock.Unlock()
-	if _, exists := p.activeSockets[protocol]; exists {
-		p.activeSockets[protocol][port] = socket
+
+	if _, exists := p.portsMap[*localPort]; exists {
+		return fmt.Errorf("error try to open socket for svc: %s/%s on port: %v again", svc.Namespace, svc.Name, port)
 	} else {
-		p.activeSockets[protocol] = map[int32]activeSocket{
-			port: socket,
+		closeable, err := utilnet.ListenPortOpener.OpenLocalPort(localPort)
+		if err != nil {
+			p.emitPortClaimEvent(svc, port, err)
+			return err
 		}
+		p.portsMap[*localPort] = closeable
 	}
 	return nil
 }
 
-func (p *portClaimWatcher) close(port int32, protocol kapi.Protocol, svc *kapi.Service) error {
+func (p *portClaimWatcher) close(desc string, ip string, port int32, protocol kapi.Protocol, svc *kapi.Service) error {
+	klog.V(5).Infof("Closing socket claimed for service: %s/%s and port: %v", svc.Namespace, svc.Name, port)
+
+	var localPort *utilnet.LocalPort
+	var err error
+	if protocol == kapi.ProtocolTCP {
+		localPort, err = utilnet.NewLocalPort(desc, ip, "", int(port), utilnet.TCP)
+	} else if protocol == kapi.ProtocolUDP {
+		localPort, err = utilnet.NewLocalPort(desc, ip, "", int(port), utilnet.TCP)
```

Comment:
You need to change `utilnet.TCP` -> `utilnet.UDP` here

Reply:
Damn, good catch. I'm surprised upstream CI doesn't have a test for this!

Reply:
fixed, a unittest added